Duplicate key error while pushing to cache

This happened when pushing to my attic cache from CI.
Since multiple machines are being built at the same time, it is possible that more than one client is pushing to the cache, so this might be a race condition somewhere.

```
May 08 04:07:40 atticd[1109]: Attic Server 0.1.0 (release)
May 08 04:07:40 atticd[1109]: Running migrations...
May 08 04:07:40 atticd[1109]: Starting API server...
May 08 04:07:40 atticd[1109]: Listening on [::]:8004...
May 11 12:27:04 atticd[1109]: 2024-05-11T11:27:04.604136Z ERROR attic_server::error: Database error: Query Error: error returned from database: duplicate key value violates unique constraint "idx-object-cache-hash"
May 11 12:27:04 atticd[1109]:    0: attic_server::api::v1::upload_path::upload_path
May 11 12:27:04 atticd[1109]:              at server/src/api/v1/upload_path.rs:117
May 11 12:27:04 atticd[1109]:    1: tower_http::trace::make_span::request
May 11 12:27:04 atticd[1109]:            with method=PUT uri=/_api/v1/upload-path version=HTTP/1.1
May 11 12:27:04 atticd[1109]:              at /nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/tower-http-0.4.4/src/trace/make_span.rs:109
May 11 12:27:04 atticd[1109]:    2: tokio::task::runtime.spawn
May 11 12:27:04 atticd[1109]:            with kind=task task.name= task.id=7642488 loc.file="/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/hyper-0.14.27/src/common/exec.rs" loc.line=49 loc.col=21
May 11 12:27:04 atticd[1109]:              at /nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-vendor-cargo-deps/c19b7c6f923b580ac259164a89f2577984ad5ab09ee9d583b888f934adbbe8d0/tokio-1.34.0/src/util/trace.rs:17
May 11 12:27:04 atticd[1109]: 2024-05-11T11:27:04.651515Z ERROR tower_http::trace::on_failure: response failed classification=Status code: 500 Internal Server Error latency=16210 ms
```

From client log:
```
❌ 29dik9ai6w9ss52zqsz8d756brg7ndvq-hwdb.bin: InternalServerError: The server encountered an internal error or misconfiguration.
// ...
 Error: InternalServerError: The server encountered an internal error or misconfiguration.
Error: Process completed with exit code 1.
```

Attic version (server and client): `4dbdbee45728d8ce5788db6461aaaa89d98081f0`
Push command arguments: `-j 2`
CI logs (might not be available for that long): https://github.com/diogotcorreia/dotfiles/actions/runs/9043154351/job/24850756629

--- a/server/src/api/v1/upload_path.rs
+++ b/server/src/api/v1/upload_path.rs
@@ -46,7 +46,7 @@ use crate::database::entity::cache;
 use crate::database::entity::chunk::{self, ChunkState, Entity as Chunk};
 use crate::database::entity::chunkref::{self, Entity as ChunkRef};
 use crate::database::entity::nar::{self, Entity as Nar, NarState};
-use crate::database::entity::object::{self, Entity as Object};
+use crate::database::entity::object::{self, Entity as Object, InsertExt};
 use crate::database::entity::Json as DbJson;
 use crate::database::{AtticDatabase, ChunkGuard, NarGuard};
 
@@ -257,12 +257,6 @@ async fn upload_path_dedup(
         .map_err(ServerError::database_error)?;
 
     // Create a mapping granting the local cache access to the NAR
-    Object::delete_many()
-        .filter(object::Column::CacheId.eq(cache.id))
-        .filter(object::Column::StorePathHash.eq(upload_info.store_path_hash.to_string()))
-        .exec(&txn)
-        .await
-        .map_err(ServerError::database_error)?;
     Object::insert({
         let mut new_object = upload_info.to_active_model();
         new_object.cache_id = Set(cache.id);
@@ -271,6 +265,7 @@ async fn upload_path_dedup(
         new_object.created_by = Set(username);
         new_object
     })
+    .on_conflict_do_update()
     .exec(&txn)
     .await
     .map_err(ServerError::database_error)?;
@@ -487,12 +482,6 @@ async fn upload_path_new_chunked(
     .map_err(ServerError::database_error)?;
 
     // Create a mapping granting the local cache access to the NAR
-    Object::delete_many()
-        .filter(object::Column::CacheId.eq(cache.id))
-        .filter(object::Column::StorePathHash.eq(upload_info.store_path_hash.to_string()))
-        .exec(&txn)
-        .await
-        .map_err(ServerError::database_error)?;
     Object::insert({
         let mut new_object = upload_info.to_active_model();
         new_object.cache_id = Set(cache.id);
@@ -501,6 +490,7 @@ async fn upload_path_new_chunked(
         new_object.created_by = Set(username);
         new_object
     })
+    .on_conflict_do_update()
     .exec(&txn)
     .await
     .map_err(ServerError::database_error)?;
@@ -594,12 +584,6 @@ async fn upload_path_new_unchunked(
     .map_err(ServerError::database_error)?;
 
     // Create a mapping granting the local cache access to the NAR
-    Object::delete_many()
-        .filter(object::Column::CacheId.eq(cache.id))
-        .filter(object::Column::StorePathHash.eq(upload_info.store_path_hash.to_string()))
-        .exec(&txn)
-        .await
-        .map_err(ServerError::database_error)?;
     Object::insert({
         let mut new_object = upload_info.to_active_model();
         new_object.cache_id = Set(cache.id);
@@ -608,6 +592,7 @@ async fn upload_path_new_unchunked(
         new_object.created_by = Set(username);
         new_object
     })
+    .on_conflict_do_update()
     .exec(&txn)
     .await
     .map_err(ServerError::database_error)?;

--- a/server/src/database/entity/object.rs
+++ b/server/src/database/entity/object.rs
@@ -6,6 +6,8 @@ use std::path::PathBuf;
 use std::str::FromStr;
 
 use sea_orm::entity::prelude::*;
+use sea_orm::sea_query::OnConflict;
+use sea_orm::Insert;
 
 use super::nar::NarModel;
 use super::Json;
@@ -14,6 +16,10 @@ use crate::narinfo::{Compression, NarInfo};
 use attic::hash::Hash;
 
 pub type ObjectModel = Model;
+
+pub trait InsertExt {
+    fn on_conflict_do_update(self) -> Self;
+}
 
 /// An object in a binary cache.
 #[derive(Debug, Clone, PartialEq, Eq, DeriveEntityModel)]
@@ -85,6 +91,27 @@ pub enum Relation {
         to = "super::nar::Column::Id"
     )]
     Nar,
+}
+
+impl InsertExt for Insert<ActiveModel> {
+    fn on_conflict_do_update(self) -> Self {
+        self.on_conflict(
+            OnConflict::columns([Column::CacheId, Column::StorePathHash])
+                .update_columns([
+                    Column::NarId,
+                    Column::StorePath,
+                    Column::References,
+                    Column::System,
+                    Column::Deriver,
+                    Column::Sigs,
+                    Column::Ca,
+                    Column::CreatedAt,
+                    Column::LastAccessedAt,
+                    Column::CreatedBy,
+                ])
+                .to_owned(),
+        )
+    }
 }
 
 impl Model {

